File: examples/linkedlist/linkedlist.cpp
#include <iostream>

using namespace std;

struct node{
	node(int x){
		data=x;
		this->next=NULL;
	}
	int data;
	node* next;
};

void addNode(struct node *h,int x){
    cout << "Add node" << endl;
    node* cur = h;
    struct node* n = new node(x);
    while(cur){
        if(cur->next==NULL){
            cur->next=n;
            break;
        }
        cur=cur->next;
    }
}

void insertFront(struct node** h,int x){
    cout << "Insert front" << endl;
    struct node* n = new node(x);
    n->next = *h;
    *h = n;
}

void reverse(struct node **h){
    cout << "Reverse" << endl;
    node *cur = *h;
    node *prev = NULL;
    while(cur){
        cout << cur->data << endl;
        //cur->next=prev;
        node *temp = cur;
        //prev=cur;
        cur=cur->next;
        temp->next=prev;
        prev=temp;
    }
    *h = prev;
}

struct node* searchNode(struct node *h,int x){
    node *cur = h;
    while(cur){
        if(cur->data==x){
            return cur;
        }
        cur=cur->next;
    }
    return NULL;
}

bool deleteNode(struct node *h,int data){
    node* cur = h;
    node *x = searchNode(cur,data);
    if(x==NULL){
        return false;
    }
    node *prev=NULL;
    while(cur!=x){
        prev=cur;
        cur=cur->next;
    }
    prev->next=cur->next;
    return true;
}

void display(struct node* h){
    cout << "display" << endl;
    node *l = h;
    while(l){
        cout << l->data << endl;
        l=l->next;
    }
}

int main(){
	struct node* head = new node(50);
	addNode(head,23);
	display(head);
	insertFront(&head,34);
	display(head);
	reverse(&head);
	display(head);
	node *x = searchNode(head,50);
	cout << x->data << endl;
	if(deleteNode(head,50)){
        cout << "Deleted" << endl;
	}
	else{
        cout << "Not" << endl;
	}
	display(head);
	return 0;
}

File: examples/linkedlist/ruby/linked_list.rb
require './node'

class LinkedList

  attr_accessor :head

  def add_node value
    if @head
      new_node = Node.new(value, nil)
      current_node = @head
      last_node = nil

      while current_node
        last_node = current_node
        current_node = current_node.next
      end

      last_node.next = new_node
    else
      new_node = Node.new(value, nil)
      @head = new_node
    end
  end

  def remove_node value
    current_node = @head
    node_to_be_deleted = nil
    prev_node = nil

    while current_node
      if current_node.value == value
        node_to_be_deleted = current_node
        break
      end

      prev_node = current_node
      current_node = current_node.next
    end

    if node_to_be_deleted
      if prev_node.nil?
        @head = node_to_be_deleted.next
      else
        prev_node.next = node_to_be_deleted.next
      end
    else
      false
    end
  end

  def is_present? value
    current_node = @head

    is_present = false
    while current_node
      if current_node.value == value
        is_present = true
        break
      end
      current_node = current_node.next
    end
    is_present
  end

  def reverse!
    current_node = @head
    previous_node = nil

    while current_node
      temp = current_node.next
      current_node.next = previous_node
      previous_node = current_node
      current_node = temp
    end

    @head = previous_node
  end

  def print
    current_node = @head

    while current_node
      puts current_node.value
      current_node = current_node.next
    end
  end

  def insert_in_front value
    if is_present? value
      return false
    end

    new_node = Node.new(value, @head)
    @head = new_node
  end

end
